src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -268,59 +265,74 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
             }
         }
 
-        private void InitializeConstructorParameters(Dictionary<string, JsonPropertyInfo> propertyCache, ConstructorInfo constructorInfo)
+        private class ParameterLookupEntry
+        {
+            public ParameterLookupEntry(JsonPropertyInfo jsonPropertyInfo)
+            {
+                JsonPropertyInfo = jsonPropertyInfo;
+            }
+
+            public JsonPropertyInfo JsonPropertyInfo { get; }
+            public JsonPropertyInfo? DuplicateJsonPropertyInfo { get; set; } = null!;
+        }
+
+        private void InitializeConstructorParameters(ConstructorInfo constructorInfo)
         {
             ParameterInfo[] parameters = constructorInfo!.GetParameters();
```

Comment:
Unrelated to your change, this is non-nullable, why do you need the forgiveness operator?

Reply:
It isn't needed. Likely due to refactoring over time; unfortunately there is no warning when ! if used when not necessary.

Reply:
Removed this one case in latest commit